cpp/density/SphereVoxelization.cc

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2010-2019 The Regents of the University of Michigan
+// This file is from the freud project, released under the BSD 3-Clause License.
+
+#include <cmath>
+#include <stdexcept>
+#include <tbb/tbb.h>
+
+#include "SphereVoxelization.h"
+
+/*! \file SphereVoxelization.cc
+    \brief Routines for computing voxelized densities from spheres centered at points.
+*/
+
+namespace freud { namespace density {
+
+SphereVoxelization::SphereVoxelization(vec3<unsigned int> width, float r_max)
+    : m_box(box::Box()), m_width(width), m_r_max(r_max)
+{
+    if (r_max <= 0.0f)
+        throw std::invalid_argument("Sphere Voxelization requires r_max to be positive.");
+}
+
+//! Get a reference to the last computed voxels
+const util::ManagedArray<unsigned int>& SphereVoxelization::getVoxels() const
+{
+    return m_voxels_array;
+}
+
+//! Get width.
+vec3<unsigned int> SphereVoxelization::getWidth()
+{
+    return m_width;
+}
+
+//! internal
+/*! \brief Function to compute the voxels array
+ */
+void SphereVoxelization::compute(const freud::locality::NeighborQuery* nq)
+{
+    auto box = nq->getBox();
+    auto n_points = nq->getNPoints();
+    m_box = box;
+
+    vec3<unsigned int> width(m_width);
+    if (box.is2D())
+    {
+        width.z = 1;
+    }
+    m_voxels_array.prepare({width.x, width.y, width.z});
+    util::ThreadStorage<unsigned int> local_bin_counts({width.x, width.y, width.z});
+
+    // set up some constants first
+    const float lx = m_box.getLx();
+    const float ly = m_box.getLy();
+    const float lz = m_box.getLz();
+    const vec3<bool> periodic = m_box.getPeriodic();
+
+    const float grid_size_x = lx / m_width.x;
+    const float grid_size_y = ly / m_width.y;
+    const float grid_size_z = m_box.is2D() ? 0 : lz / m_width.z;
+
+    // Find the number of bins within r_max
+    const int bin_cut_x = int(m_r_max / grid_size_x);
+    const int bin_cut_y = int(m_r_max / grid_size_y);
+    const int bin_cut_z = m_box.is2D() ? 0 : int(m_r_max / grid_size_z);
+    const float r_max_sq = m_r_max * m_r_max;
+
+    util::forLoopWrapper(0, n_points, [&](size_t begin, size_t end) {
+        // for each reference point
+        for (size_t idx = begin; idx < end; ++idx)
+        {
+            const vec3<float> point = (*nq)[idx];
+            // Find which bin the particle is in
+            const int bin_x = int((point.x + lx / 2.0f) / grid_size_x);
+            const int bin_y = int((point.y + ly / 2.0f) / grid_size_y);
+            // In 2D, only loop over the z=0 plane
+            const int bin_z = m_box.is2D() ? 0 : int((point.z + lz / 2.0f) / grid_size_z);
+
+            // Reject bins that are outside the box in aperiodic directions
+            // Only evaluate over bins that are within the cutoff
+            for (int k = bin_z - bin_cut_z; k <= bin_z + bin_cut_z; k++)
+            {
+                if (!periodic.z && (k < 0 || k >= int(m_width.z)))
```

Comment:
I would like to see this conditional logic structure replicated in `GaussianDensity`. It should be faster to evaluate and reject the inner loops in advance.

Reply:
I concur